### ISSUE TITLE
Some test failures should be nonblocking.

### ISSUE THREAD

**Issue by polina-c** [flutter/flutter#140466]
Flutter CI is doing randomization testing with seed of today date.
There are no reasons for this testing to be blocking for flutter master branch.

**Comment by polina-c** [comment]
@godofredoc @goderbauer @keyonghan @jacob314 

Thoughts?

**Comment by polina-c** [comment]
+ @zanderso

**Comment by zanderso** [comment]
Test shuffling catches real test hermeticity errors. If test shuffling isn't blocking, then IIUC we would have to run tests twice: once without shuffling, which would be blocking, and once with shuffling, which would not be blocking.

**Comment by polina-c** [comment]
> Test shuffling catches real test hermeticity errors. If test shuffling isn't blocking, then IIUC we would have to run tests twice: once without shuffling, which would be blocking, and once with shuffling, which would not be blocking.

I thought we are doing it anyway: two times, with shuffling and without shuffling. My suggestion is to make shuffling non-blocking. 

Yes, shuffling catches real test hermeticity errors, that is very good thing. The problem is that an error created by one engineer blocks other engineers, that impacts velocity of the team.

Are there reasons to make shuffled testing blocking?

**Comment by christopherfujino** [comment]
> I thought we are doing it anyway: two times, with shuffling and without shuffling. Am I wrong? My suggestion is to make shuffling non-blocking.

For package:test tests, we only run them once, with shuffling.

> Yes, shuffling catches real test hermeticity errors, that is very good thing. The problem is that an error created by one engineer blocks other engineers, that impacts velocity of the team.

I think if we had a reliable system to run them in a non-blocking fashion but still file issues and/or TODOs in the code that some team needs to be responsible for fixing the test, I think that would be a productivity win over what we have now, which is basically that when the daily shuffle causes failures, the framework gardener needs to investigate the failure, deduce from the failure message and the test code that the failure is due to the random shuffle seed, and create a PR to mark the test library no-shuffle.

**Comment by polina-c** [comment]
If framework gardener is doing this anyway, how about making it their responsibility to check the status of the non-blocking bot and file issues and/or TODOs?

It will not change anything for the gardener, but will save the team from being blocked.

**Comment by christopherfujino** [comment]
> It will not change anything for the gardener, but will save the team from being blocked.

It will change the places the gardener needs to check from 1 to 2.

**Comment by polina-c** [comment]
> It will change the places the gardener needs to check from 1 to 2.

Both bots are on [dashboard](https://flutter-dashboard.appspot.com/#/build). What is the second place?

**Comment by christopherfujino** [comment]
> > It will change the places the gardener needs to check from 1 to 2.
> 
> Both bots are on [dashboard](https://flutter-dashboard.appspot.com/#/build). What is the second place?

We don't have any current infrastructure to run tests twice, both with shuffle and without. I'm assuming that you are proposing making the existing tests run without shuffle, and setting a separate, asynchronous infrastructure to run tests WITH a shuffle in a non-blocking way. The monitoring for this would be the second place.

**Comment by polina-c** [comment]
> > > It will change the places the gardener needs to check from 1 to 2.
> > 
> > 
> > Both bots are on [dashboard](https://flutter-dashboard.appspot.com/#/build). What is the second place?
> 
> We don't have any current infrastructure to run tests twice, both with shuffle and without. I'm assuming that you are proposing making the existing tests run without shuffle, and setting a separate, asynchronous infrastructure to run tests WITH a shuffle in a non-blocking way. The monitoring for this would be the second place.

Thanks for phrasing this. Yes, this is what i suggesting. 
This second place does not require separate infrastructure. It will be also flutter dashboard. There are two parts: non-blocking at the left and blocking at the right.
To make a job non-blocking, you need to set `bringup: true` in [`.ci.yaml`](https://github.com/flutter/flutter/blob/0503c4471834017da79f5b34090b1a37a2b04279/.ci.yaml#L2036) for it.

**Comment by goderbauer** [comment]
Currently, the job of the gardener is very simple: Keep the tree green.

It is well established, that tests with `bringup: true` (those that show up as outlined squares on the dashboard) can be ignored - and in fact this is encoded in our infra as the tree will not go red on those.

The suggestion would make the gardner's job more complicated as they would have to monitor individual tests on the dashbaord that - without any additional infra work - are indistinguishable from tests that can be safely ignored. I don't think this will scale well and we are likely going to miss issues due to accidental human errors or gardeners just not knowing that this is there additional responsibility. For keeping the tree green the entire team is keeping the gardener honest (and incentivized to help the gardener) because the team will be blocked from submitting their PRs while the tree is red (see the tree-status chat on discord). No such "oversight" exists for `bringup: true` tests. This is compounded by the fact that most randomization issues _seemingly_ "fix themselves" when the seed changes the next day. Overall, I am afraid that with the suggested approach there is very little incentive to actually record (and then fix) the underlying issue and we will accumulate a lot of test hermeticity problems - that will potentially lead to larger problems down the road.

tl;dr: I believe the gardener rotation is so successful because the mission as a gardener is dead simple: keep the tree green. Making the gardener's job more complicated is likely not in our interest of keeping the tree as green as possible, which will hurt team velocity and productivity.

**Comment by polina-c** [comment]
thanks for discussion. closing

**Comment by polina-c** [comment]
How about introducing one more color: non blocking yellow. It will serve both goals:

1. Gardener's job will still be to keep tree green
2. Tech debt like hermiticity of tests and deprecations will be handled without blocking the team

@jonahwilliams @matanlurey

**Comment by goderbauer** [comment]
> High priority tech debt like hermiticity of tests and deprecations will be handled without blocking the team

Who will be handling them? What is the suggested process?

**Comment by polina-c** [comment]
> Who will be handling them? What is the suggested process?

They will be handled the same way as today:
1. Gardener notices the tree is not green (it is yellow)
2. Gardener either fixes the issues or submits PR to ignore them with TODO

The only difference will be in that the issues will be not-blocking

**Comment by polina-c** [comment]
Closing as duplicate of https://github.com/flutter/flutter/issues/143344
